### PR TITLE
Surface error "missing targets column" early

### PR DESCRIPTION
**Pull request by annzhang-db**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/annzhang-db/mlflow/pull/10020?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/10020/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 10020
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

- add defaults to builtin metrics for targets so that we can surface errors for missing the 'targets' column for metrics that require them early
- eval_fn that don't require targets column must have a default
- pass in default for parameters that have a default

### How is this PR tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests

Now we can have this error message surfaced in _get_args_for_metrics:

Screenshot 2023-10-19 at 5 47 14 PM

instead of the following when we actually call eval_fn:
Screenshot 2023-10-19 at 11 19 59 AM

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/metrics/metric_definitions.py [hunk 1]
```diff
@@ -254,47 +254,47 @@ def _rougeLsum_eval_fn(predictions, targets, metrics):
...
+def _r2_score_eval_fn(predictions, targets=None, metrics=None, sample_weight=None):
     if targets is not None and len(targets) != 0:
         from sklearn.metrics import r2_score
 
         r2 = r2_score(targets, predictions, sample_weight=sample_weight)
         return MetricValue(aggregate_results={"r2_score": r2})
 
 
-def _max_error_eval_fn(predictions, targets, metrics):
+def _max_error_eval_fn(predictions, targets=None, metrics=None):
```
Comment (sunishsheth2009) [inline mlflow/metrics/metric_definitions.py:289]: Why do we need to explicitly set this to None? If users create a custom_metric and not set this to None. Would the error handling work? 

Also in some of the metrics, targets is actually required since we would do a no-op if target is not passed in right? 🤔

File: mlflow/models/evaluation/default_evaluator.py [hunk 2]
```diff
@@ -1423,29 +1428,27 @@ def _check_args(self, metrics, eval_df):
...
-                f"Input Columns: {input_columns}\n"
-                f"Output Columns: {output_columns}\n"
-                "Note that this does not include the output column: "
-                f"'{output_column_name}'\n\n"
-                f"To resolve this issue, you may want to map the missing column to an "
-                "existing column using the following configuration:\n"
-                f"evaluator_config={{'col_mapping': {{'<missing column name>': "
-                "'<existing column name>'}}\n"
-            )
+            error_messages = [
```
Comment (sunishsheth2009) [inline mlflow/models/evaluation/default_evaluator.py:1436]: Can we add a quick test case for this `error_messages` being rendered correctly since it seems like a complicated error message

File: mlflow/metrics/metric_definitions.py (hunk 1 above)
Comment (annzhang-db) [reply mlflow/metrics/metric_definitions.py:289]: Yes, we want to do a no-op if targets is None and skip logging that metric (if there is no default it will fail the evaluation saying that 'metric_name' requires parameter 'targets')

Comment (annzhang-db) [reply mlflow/metrics/metric_definitions.py:289]: The reason for this was that _get_args_for_metrics will surface errors for other required columns (such as 'context') but it will not do the same for targets right now - missing 'targets' column (for instance for answer similarity) won't be surfaced until we actually call eval_fn which will raise an ambiguous error since we don't know if targets is "required" or not until then

File: mlflow/models/evaluation/default_evaluator.py (hunk 2 above)
Comment (annzhang-db) [reply mlflow/models/evaluation/default_evaluator.py:1447]: modified existing test

**Review by prithvikannan — APPROVED** [review]
LGTM, discussed offline and this makes sense to me.

File: mlflow/models/evaluation/default_evaluator.py [hunk 3]
```diff
@@ -1194,6 +1197,8 @@ def _get_args_for_metrics(self, extra_metric, eval_df):
                         eval_fn_args.append(self.evaluator_config.get(column))
                     elif param.default == inspect.Parameter.empty:
                         params_not_found.append(param_name)
+                    else:
+                        eval_fn_args.append(param.default)
```
Comment (sunishsheth2009) [inline mlflow/models/evaluation/default_evaluator.py:1201]: Do we need to do this? Isn't this already done by default?

**Review by sunishsheth2009 — APPROVED** [review]

File: mlflow/models/evaluation/default_evaluator.py (hunk 3 above)
Comment (annzhang-db) [reply mlflow/models/evaluation/default_evaluator.py:1201]: In the case that the eval_fn has parameters (predictions, targets=None, metrics=None) for example - if targets is not provided (but has default value) then the eval_fn_args becomes [predictions, metrics] when we want it to be [predictions, None, metrics] and metrics actually gets passed into targets (with metrics taking its own default value)

Comment (annzhang-db) [reply mlflow/models/evaluation/default_evaluator.py:1201]: LMK if this doesn't make sense!
